### DISCUSSION TITLE
How to make an extentable rope?

### DISCUSSION THREAD

**Discussion by lucasmaffazioli** [appsinacup/godot-rapier-physics#316] [Q&A]
I’m trying to do this for literally weeks/months now, but every time I think I’m close to a solution, it fails. This is my current game, you can see a rope made of several rigidbodies.

https://github.com/user-attachments/assets/3fe21463-026e-4efa-a6ae-f6d2cde21b0d

So my frog player shoots a rope that glues to the surface and to him. My frog is a rigidbody, because I want him to be effected by the physics engine and other characters. I also created the rope to be a rigidbody, so it can interact with the world and other characters as well. 

**The challenge: To shorten and extend the rope length. Seems simple enough, but:**

1 changing the length of the rigidbodies don’t do anything because the pinjoint stays in the same place,

2 changing the connecting place of the pinjoint is also not possible. I tried with some hacks but nothing good happened with it

3 I tried connecting each pinjoint to a static body that was connected to the rigidbody and when extending I would just move the static body outward, and although I had some success while testing this, it just became too unstable with the actual rope, everything disconnects

Any other suggestions would be greatly appreciated, thanks!

**Comment by Ughuuu** [comment]
Can you delete one of the rigidbodies and move the pin joint A from the previous rigidbody to the static body connection point? Would it move it by force to the new point in that case?

Eg:
StaticBody -> RigidBodyA -> RigidBodyB ->...
delete RigidBodyA
Set RigidBodyB pin join A to StaticBody instead of what it was previously.

**Reply by lucasmaffazioli** [reply]
Yes that's the 3. point, it just becomes utterly unstable:

https://github.com/user-attachments/assets/a5a4b2aa-762f-44a2-b6fa-3bb9e386ae95


Maybe what I need is something like [RopeJoint](https://github.com/dimforge/rapier/pull/415)? The example on this pull without the RopeJoint behaves exactly how it does on my tests. I think this could potentially solve some issues. I don't think it's implemented on this plugin though?

**Comment by Ughuuu** [comment]
Do you have any examples I can look at, an example project I can test this? The rope joint is not exposed rn at all through godot rapier afaik.
It looks like either the rigidbody2D player is very heavy, or something really weird is happening.

**Reply by lucasmaffazioli** [reply]
Sure, I'll send you the somewhat working solution I designed, but becomes very unstable like in the video when the rope connects to a staticBody

 in this project theres these scenes, in both of them the rope is connected through a series of rigidbodies with static bodies to serve as anchor:

- test_rope: rope connected to a rigid body
- test_rope2: rope connected to a static body (here is mayhem)


Controls: AWD move player, F,G, pull or push anchors to shorten/extend rope


https://easyupload.io/5wbn1q


obs./extra info: In Godot's 2D engine, the visual result is slightly different than in test_rope2, but the actual physical result is the same

**Comment by lucasmaffazioli** [comment]
I ended up being able to use DampedSpringJoint2D, realised that I was not using length and initial_length correctly, basically I must change both at the same time for the expanding and contraction to happen. Thanks again for porting Rapier!

**Comment by Ughuuu** [comment]
Nice, show a video of final product, am curious.

**Reply by lucasmaffazioli** [reply]
https://github.com/user-attachments/assets/af84118f-a25c-4741-a1bb-9e2203baff21

Here it is the very basic of it, but of course this "rope" doesn't collide with objects or anything, but probably is good enough for me at the moment
